src/DotLiquid/Context.cs

```diff
@@ -561,73 +536,63 @@ private object Variable(string markup, bool notifyNotFound)
+                value = keyValuePair.Value;
+            else if (obj is IDictionary dictionaryObj && dictionaryObj.Contains(key))
                 value = dictionaryObj[key];
-            }
-            else if (obj is IList listObj)
-            { 
+            // Resolve #350/#417, add support for rendering of a nested ExpandoObject
+            else if (obj is IDictionary<string, object> dictionaryObject && dictionaryObject.ContainsKey(key.ToString()))
+                value = dictionaryObject[key.ToString()];
+            else if (obj is IList listObj && (key is int || key is long))
                 value = listObj[Convert.ToInt32(key)];
-            }
-            else if (TypeUtility.IsAnonymousType(obj.GetType()))
-            { 
+            else if (TypeUtility.IsAnonymousType(obj.GetType()) && obj.GetType().GetRuntimeProperty((string)key) != null)
                 value = obj.GetType().GetRuntimeProperty((string)key).GetValue(obj, null);
-            }
-            else if (obj is IIndexable indexableObj)
-            { 
+            else if (obj is IIndexable indexableObj && indexableObj.ContainsKey(key))
                 value = indexableObj[key];
-            }
-            else
-            { 
-                throw new NotSupportedException();
+            else if (obj is IEnumerable enumerableObject && key is string enumerableKey)
+            {
+                // Special cases: If no key with the same name was found we interpret size|first|last
+                // as commands and call them on the current IEnumerable object
+                var castCollection = enumerableObject.Cast<object>();
+                if (Template.NamingConvention.OperatorEquals(enumerableKey, "size"))
+                    value = castCollection.Count();
+                else if (Template.NamingConvention.OperatorEquals(enumerableKey, "first"))
+                    value = castCollection.FirstOrDefault();
+                else if (Template.NamingConvention.OperatorEquals(enumerableKey, "last"))
+                    value = castCollection.LastOrDefault();
             }
 
             if (value is Proc procValue)
             {
                 object newValue = procValue.Invoke(this);
                 if (obj is IDictionary dicObj)
-                {
                     dicObj[key] = newValue;
-                }
                 else if (obj is IList listObj)
-                {
                     listObj[Convert.ToInt32(key)] = newValue;
-                }
                 else if (TypeUtility.IsAnonymousType(obj.GetType()))
-                { 
                     obj.GetType().GetRuntimeProperty((string)key).SetValue(obj, newValue, null);
-                }
                 else
-                { 
                     throw new NotSupportedException();
-                }
-                return newValue;
+
+                value = newValue;
             }
 
-            return value;
+            return value != null;

```

Comment:
Value can be explicitly nuull and should return true.

Reply:
Done